executor/common_linux.h

```diff
@@ -181,30 +182,71 @@ static int netlink_send_ext(struct nlmsg* nlmsg, int sock,
 	memset(&addr, 0, sizeof(addr));
 	addr.nl_family = AF_NETLINK;
 	unsigned n = sendto(sock, nlmsg->buf, hdr->nlmsg_len, 0, (struct sockaddr*)&addr, sizeof(addr));
-	if (n != hdr->nlmsg_len)
-		fail("short netlink write: %d/%d", n, hdr->nlmsg_len);
+	if (n != hdr->nlmsg_len) {
+		debug("short netlink write: %d/%d", n, hdr->nlmsg_len);
```

Comment:
Can this (and subsequent ones) legitimately fail now?
We used `fail` here for a reason. This is not exactly normal production C code. Because this is fuzzer-driven lots of errors are silently ignored (and some callers of this function don't even bother checking return code). So if something fails unexpectedly (a bug in executor code), it won't be noticed. We will just get some drop in coverage which nobody will notice.
If there is a fail, it will be noticed as it will crash whole VMs. And since we had fail here for some time now and proved to not fail all the time (one of few bits here that one can really rely on now), it's pity to start giving up ground.

Reply:
If fault injections are enabled, VMs begin to constantly crash as `lost connection to test machine`. So we either need to avoid the use of these netlink helpers for pseudo syscalls (which will bloat code size), or make them fail gracefully. 

I re-implemented this patch, now it is possible to specify whether the function should handle an error gracefully or not. The implementation is a bit hacky in order to keep the function short and avoid duplication.

Reply:
Ouch. Good find. Frankly, I never considered such a possibility. But it looks very realistic.

However, the current approach you took does not look very scalable. This problem also seems to affect more functions. syz_init_net_socket uses fail and it can perfectly fail with fault injection. read_tun -- the same (maybe we even observed this, there is some strange logic for errno). And I assume more functions are subject to this. With your approach we would need to carefully examine code, explicitly opt-in each function and then carefully use the graceful flag. Any wrong value for graceful will lead to the same problem, but as we see it's quite hard to notice, so most likely these bugs will stay there for a very long time.

I wonder if we could change fail instead. If fault injection is enabled, it could exit with 0 status like exitf. This will solve the problem once and for all, and will also be a good place for a comment.
To check if fault injection is enabled right now for the current thread, we will need to add some __thread variable somewhere. We already have current_cover, maybe we should have current_thread instead with all the data regarding current thread. I can't say what exactly is the right way to organize it, it needs some experimentation to find a nice solution that works.

Reply:
That would indeed cover all benign failures. But do we really want to terminate execution in such cases? Technically, nothing prevents the fuzzer from running other syscalls there. And it might even be desirable to continue execution as further syscalls might leverage the mess, which was possibly created by failure injection, and hit more intricate kernel bugs. Exiting can cause the kernel to clean things up and some kernel bugs may just go unnoticed.

Reply:
My reasoning was two-fold:
1. I don't see a nice way to achieve this. If we make fail return in some cases, we will need to change all call sites to handle return. And it will be somewhat confusing semantics for a fail function, and it will be exercised in very rare cases, so developers won't even test these scenarios. People also get used to fail being no-return, so we will sure get bugs in future where we don't handle return case. And for e.g. syz_init_net_socket there is no way out, we need to restore the test namespace.
2. I don't think it gives too much interesting coverage. It only affects pseudo-syscalls and it can achieve any new coverage it gets here with other means (normal syscalls). It will only make few programs where we inject failures into particular points in particular pseudo-syscalls exit early, which seems fine. Programs exit prematurely all the time (e.g. if the test contains exit_group).

Reply:
Thank you for the detailed explanation.

I implemented it as you suggested - now there's a thread-local variable that is used to access information about the thread itself. I tested the patch locally, now syzkaller does not generate regular `lost connection to test machine`  crashes when fault injection is enabled.